src/Codegen/Codegen.hs

```diff
@@ -586,72 +586,92 @@ genPrim I.Return [e] _ = do
+  ((lhsVal, rhsVal), stms) <-
+    first (bimap unmarshal unmarshal) <$> genBinop lhs rhs
+  return (marshal [cexp|$exp:lhsVal / $exp:rhsVal|], stms)
 genPrimOp I.PrimMod [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
-  return ([cexp|$exp:lhsVal % $exp:rhsVal|], stms)
+  ((lhsVal, rhsVal), stms) <-
+    first (bimap unmarshal unmarshal) <$> genBinop lhs rhs
+  return (marshal [cexp|$exp:lhsVal % $exp:rhsVal|], stms)
 genPrimOp I.PrimNeg [opr] _ = do
-  (val, stms) <- genExpr opr
-  return ([cexp|- $exp:val|], stms)
+  (val, stms) <- first unmarshal <$> genExpr opr
+  return (marshal [cexp|- $exp:val|], stms)
 genPrimOp I.PrimBitAnd [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
+  ((lhsVal, rhsVal), stms) <- genBinop lhs rhs
   return ([cexp|$exp:lhsVal & $exp:rhsVal|], stms)
 genPrimOp I.PrimBitOr [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
+  ((lhsVal, rhsVal), stms) <- genBinop lhs rhs
   return ([cexp|$exp:lhsVal | $exp:rhsVal|], stms)
 genPrimOp I.PrimBitNot [opr] _ = do
   (val, stms) <- genExpr opr
-  return ([cexp|~ $exp:val|], stms)
+  -- all integers are 31 bits + 1 tag bit, so val XOR (~1)
+  -- flips the 31 bits and keeps the tag bit 1. 
+  return ([cexp|($exp:val ^ 0xFFFFFFFE)|], stms)
 genPrimOp I.PrimEq [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
+  ((lhsVal, rhsVal), stms) <- genBinop lhs rhs
   return ([cexp|$exp:lhsVal == $exp:rhsVal|], stms)
 genPrimOp I.PrimNot [opr] _ = do
-  (val, stms) <- genExpr opr
-  return ([cexp|! $exp:val|], stms)
+  (val, stms) <- first unmarshal <$> genExpr opr
+  return (marshal [cexp|! $exp:val|], stms)
 genPrimOp I.PrimGt [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
+  ((lhsVal, rhsVal), stms) <- genBinop lhs rhs
   return ([cexp|$exp:lhsVal < $exp:rhsVal|], stms)
 genPrimOp I.PrimGe [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
+  ((lhsVal, rhsVal), stms) <- genBinop lhs rhs
   return ([cexp|$exp:lhsVal <= $exp:rhsVal|], stms)
 genPrimOp I.PrimLt [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
+  ((lhsVal, rhsVal), stms) <- genBinop lhs rhs
   return ([cexp|$exp:lhsVal > $exp:rhsVal|], stms)
 genPrimOp I.PrimLe [lhs, rhs] _ = do
-  (lhsVal, rhsVal, stms) <- genBinop lhs rhs
+  ((lhsVal, rhsVal), stms) <- genBinop lhs rhs
   return ([cexp|$exp:lhsVal >= $exp:rhsVal|], stms)
 genPrimOp _ _ _ = fail "Unsupported PrimOp or wrong number of arguments"
 
 -- | Helper for sequencing across binary operations.
-genBinop :: Expr -> Expr -> GenFn (C.Exp, C.Exp, [C.BlockItem])
+genBinop :: Expr -> Expr -> GenFn ((C.Exp, C.Exp), [C.BlockItem])
 genBinop lhs rhs = do
   (lhsVal, lhsStms) <- genExpr lhs
   (rhsVal, rhsStms) <- genExpr rhs
-  return (lhsVal, rhsVal, lhsStms ++ rhsStms)
+  return ((lhsVal, rhsVal), lhsStms ++ rhsStms)
+
+-- | Marshal a C expression evaluating to a 32 bit int
+marshal :: C.Exp -> C.Exp
+marshal e = [cexp|((((typename uint32_t) $exp:e) << 1) | 0x1)|]
```

Comment:
nit: we use `typename uint32_t` a lot of places, so perhaps this should be extracted into `Codegen.Identifiers`.

Perhaps we could introduce a `word_t`:

```
--- i/src/Codegen/Identifiers.hs
+++ w/src/Codegen/Identifiers.hs
@@ -88,6 +88,9 @@ priority = "priority"
 sv_t :: C.Type
 sv_t = [cty|struct sv|]

+word_t :: C.Type
+word_t = [cty|typename uint32_t|]
+
 -- | The conventional name of the payload field in a type-specialized SV.
 value :: CIdent
 value = "value"
```

and then replace all occurrences of `(typename uint32_t)` with that to make it clearer